### PR TITLE
Bank: history of user account operations

### PR DESCRIPTION
**Pull request by KrasimirRaikov**
History with the users operations, devided in pages of twenty.
history

### CODE REVIEW

**Comment by panayotkulchev** [comment]
PR will be reviewed today in the evening

File: src/main/java/com/clouway/bank/core/AccountHistoryRepository.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.clouway.bank.core;
+
+import java.util.List;
+
+/**
+ * @author Krasimir Raikov(raikov.krasimir@gmail.com)
+ */
+public interface AccountHistoryRepository {
+
+  List<AccountRecord> history(String userId, int offset, int limit);
```
Comment (panayotkulchev) [inline src/main/java/com/clouway/bank/core/AccountHistoryRepository.java]: `history` is nown.  `getAccountRecords()`

File: src/main/java/com/clouway/bank/core/AccountPager.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * @author Krasimir Raikov(raikov.krasimir@gmail.com)
+ */
+public interface AccountPager {
+  void calculateNext();
+
+  void calculatePrevious();
+
+  List<AccountRecord> getCurrentPage(String userId);
```
Comment (panayotkulchev) [inline src/main/java/com/clouway/bank/core/AccountPager.java]: Do you have to call `calculateNext()` befor you call `getCurrentPage()`

File: src/main/java/com/clouway/bank/core/AccountPager.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+import com.clouway.bank.core.AccountRecord;
+
+import java.util.List;
+
+/**
+ * @author Krasimir Raikov(raikov.krasimir@gmail.com)
+ */
+public interface AccountPager {
+  void calculateNext();
```
Comment (panayotkulchev) [inline src/main/java/com/clouway/bank/core/AccountPager.java]: Why you do not have 
List<AccountRecords> records = accountPager.nextPage();
Now you always coll calculate next and then getCurrentPage()...

File: src/main/java/com/clouway/bank/http/AccountHistoryPage.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+  @Override
+  public void init() throws ServletException {
+    template.loadFromFile("web/WEB-INF/pages/AccountHistory.html");
+  }
+
+  protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+    String userId = "Stanislava";
+
+    TimeZone.setDefault(TimeZone.getTimeZone("UTC"));
```
Comment (panayotkulchev) [inline src/main/java/com/clouway/bank/http/AccountHistoryPage.java]: Why is this needed

File: src/main/java/com/clouway/bank/http/AccountHistoryPage.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+/**
+ * @author Krasimir Raikov(raikov.krasimir@gmail.com)
+ */
+@WebServlet(name = "AccountHistoryPage")
+public class AccountHistoryPage extends HttpServlet {
+  private Template template;
+  private TableBuilder tableBuilder;
+  private AccountPager accountPager;
+
+  public AccountHistoryPage(Template template, TableBuilder tableBuilder, AccountPager accountPager) {
```
Comment (panayotkulchev) [inline src/main/java/com/clouway/bank/http/AccountHistoryPage.java]: If couple users are on this page are they going to avoid if first open the third page and go to forth, when the second user open the page he will not go to forth page directly

File: src/main/java/com/clouway/bank/http/AccountHistoryPage.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    this.template = template;
+    this.tableBuilder = tableBuilder;
+    this.accountPager = accountPager;
+  }
+
+  @Override
+  public void init() throws ServletException {
+    template.loadFromFile("web/WEB-INF/pages/AccountHistory.html");
```
Comment (panayotkulchev) [inline src/main/java/com/clouway/bank/http/AccountHistoryPage.java]: If there is no next page, is the nextPage button active. If so it should not.

File: src/main/java/com/clouway/bank/persistent/PersistentAccountHistoryRepository.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+ * @author Krasimir Raikov(raikov.krasimir@gmail.com)
+ */
+public class PersistentAccountHistoryRepository implements AccountHistoryRepository {
+  private ConnectionProvider connectionProvider;
+
+  public PersistentAccountHistoryRepository(ConnectionProvider connectionProvider) {
+    this.connectionProvider = connectionProvider;
+  }
+
+  private static double round(double value, int places) {
```
Comment (panayotkulchev) [inline src/main/java/com/clouway/bank/persistent/PersistentAccountHistoryRepository.java]: Why this method is static in case you use it only in this repository

File: src/main/java/com/clouway/utility/Cell.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.clouway.utility;
+
+/**
+ * @author Krasimir Raikov(raikov.krasimir@gmail.com)
+ */
+public interface Cell {
```
Comment (panayotkulchev) [inline src/main/java/com/clouway/utility/Cell.java]: Empty interface... I am getting tired of review PR that the owner do not even reviewed his code. You do this very often. Review your PR and then I will continue

File: src/main/java/com/clouway/bank/core/AccountHistoryRepository.java (hunk 1 above)
Comment (KrasimirRaikov) [inline src/main/java/com/clouway/bank/core/AccountHistoryRepository.java]: renamed it to ```getAccountRecords```

File: src/main/java/com/clouway/bank/http/AccountHistoryPage.java (hunk 6 above)
Comment (KrasimirRaikov) [inline src/main/java/com/clouway/bank/http/AccountHistoryPage.java]: It's not anymore.

File: src/main/java/com/clouway/bank/persistent/PersistentAccountHistoryRepository.java (hunk 7 above)
Comment (KrasimirRaikov) [inline src/main/java/com/clouway/bank/persistent/PersistentAccountHistoryRepository.java]: it's no longer static.

**Comment by KrasimirRaikov** [comment]
PR notes fixed, waiting review.

File: src/main/java/com/clouway/bank/core/AccountHistoryPager.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+  @Override
+  public int getNextPageNumber() {
+    return nextPage;
+  }
+
+  @Override
+  public List<AccountRecord> requestPage(String userId, int pageNumber) {
+    List<AccountRecord> records;
+    int numberOfRecords = historyRepository.countRecords(userId);
```
Comment (panayotkulchev) [inline src/main/java/com/clouway/bank/core/AccountHistoryPager.java]: to count records every time you request page is not good idea. After that you make second request to DB to get the records of the page. This can be optimized with only one request to DB. If pageSize is 10, you can load 11 records. if you get 11 records this means there is next page you can display.

Comment (KrasimirRaikov) [inline src/main/java/com/clouway/bank/core/AccountHistoryPager.java]: Did it as suggested.

**Comment by KrasimirRaikov** [comment]
PR notes are fixed.

File: src/main/java/com/clouway/bank/http/AccountHistoryPage.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    template.put("previous", String.valueOf(0));
+    template.put("previousState", "auto");
+    template.put("next", String.valueOf(1));
+    template.put("nextState", "auto");
+
+    template.put("currentPage", String.valueOf(0));
```
Comment (panayotkulchev) [inline src/main/java/com/clouway/bank/http/AccountHistoryPage.java]: Is it going to work if use:
`template.put("currentPage", "0");`

Comment (KrasimirRaikov) [inline src/main/java/com/clouway/bank/http/AccountHistoryPage.java]: It will.

**Comment by KrasimirRaikov** [comment]
Fixed a problem with the NumberFormatException, and calling too big number for the page, and the comment about getting the String value of the int.

File: src/test/java/com/clouway/bank/http/AccountHistoryPageTest.java [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+  SessionProvider sessionProvider;
+  private AccountHistoryPage historyPage;
+
+  @Before
+  public void setUp() {
+    historyPage = new AccountHistoryPage(template, tableBuilder, accountPager, sessionProvider);
+  }
+
+  @Test
+  public void displayFirstPage() throws ServletException, IOException {
```
Comment (panayotkulchev) [inline src/test/java/com/clouway/bank/http/AccountHistoryPageTest.java]: Only tell me do not do it. Could you count me some more scenarios?

Comment (KrasimirRaikov) [inline src/test/java/com/clouway/bank/http/AccountHistoryPageTest.java]: Well, I have tested displaying the pages, I have tested the disabling of the buttons, so I could test incorrect input in the request, I could also test displaying page for empty history

File: src/test/java/com/clouway/bank/AccountHistoryPagerTest.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  private AccountPager accountPager;
+  private int pageSize = 20;
+
+  @Before
+  public void setUp() {
+    accountPager = new AccountHistoryPager(pageSize, accountHistoryRepository);
+  }
+
+  @Test
+  public void calculatingSecondPage() {
```
Comment (KrasimirRaikov) [inline src/test/java/com/clouway/bank/AccountHistoryPagerTest.java]: I have added another test
